File: android/src/main/java/com/unact/yandexmapkit/YandexPedestrianSession.java
package com.unact.yandexmapkit;

import androidx.annotation.NonNull;

import com.yandex.mapkit.transport.masstransit.Session;

import io.flutter.plugin.common.BinaryMessenger;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;

public class YandexPedestrianSession implements MethodChannel.MethodCallHandler {
    private final int id;
    private final Session session;
    private final MethodChannel methodChannel;
    private final YandexPedestrian.PedestrianCloseListener closeListener;

    public YandexPedestrianSession(
            int id,
            Session session,
            BinaryMessenger messenger,
            YandexPedestrian.PedestrianCloseListener closeListener
    ) {
        this.id = id;
        this.session = session;
        this.closeListener = closeListener;

        methodChannel = new MethodChannel(messenger, "yandex_mapkit/yandex_pedestrian_session_" + id);
        methodChannel.setMethodCallHandler(this);
    }

    @Override
    public void onMethodCall(MethodCall call, @NonNull MethodChannel.Result result) {
        switch (call.method) {
            case "cancel":
                cancel();
                result.success(null);
                break;
            case "retry":
                retry(result);
                break;
            case "close":
                close();
                result.success(null);
                break;
            default:
                result.notImplemented();
                break;
        }
    }

    public void cancel() {
        session.cancel();
    }

    public void retry(MethodChannel.Result result) {
        session.retry(new YandexPedestrianListener(result));
    }

    public void close() {
        session.cancel();
        methodChannel.setMethodCallHandler(null);

        closeListener.onClose(id);
    }
}

File: ios/Classes/YandexPedestrianSession.swift
import Foundation
import YandexMapsMobile

public class YandexPedestrianSession: NSObject {
  private var id: Int
  private var session: YMKMasstransitSession
  private let methodChannel: FlutterMethodChannel!
  private var onClose: (Int) -> ()

  public required init(
    id: Int,
    session: YMKMasstransitSession,
    registrar: FlutterPluginRegistrar,
    onClose: @escaping ((Int) -> ())
  ) {
    self.id = id
    self.session = session
    self.onClose = onClose

    methodChannel = FlutterMethodChannel(
      name: "yandex_mapkit/yandex_pedestrian_session_\(id)",
      binaryMessenger: registrar.messenger()
    )

    super.init()

    weak var weakSelf = self
    self.methodChannel.setMethodCallHandler({ weakSelf?.handle($0, result: $1) })
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
    case "cancel":
      cancel()
      result(nil)
    case "retry":
      retry(result)
    case "close":
      close()
      result(nil)
    default:
      result(FlutterMethodNotImplemented)
    }
  }

  public func cancel() {
    session.cancel()
  }

  public func retry(_ result: @escaping FlutterResult) {
      session.retry(routeHandler: {(pedestrianResponse: [YMKMasstransitRoute]?, error: Error?) -> Void in
      self.handleResponse(pedestrianResponse: pedestrianResponse, error: error, result: result)
    })
  }

  public func close() {
    session.cancel()

    onClose(id)
  }

  public func handleResponse(pedestrianResponse: [YMKMasstransitRoute]?, error: Error?, result: @escaping FlutterResult) {
    if let response = pedestrianResponse {
      onSuccess(response, result)
    } else {
      onError(error!, result)
    }
  }

  private func onSuccess(_ res: [YMKMasstransitRoute], _ result: @escaping FlutterResult) {
    let routes = res.map { (route) -> [String: Any] in
      let weight = route.metadata.weight

      return [
        "geometry": route.geometry.points.map {
          (point) -> [String: Any] in Utils.pointToJson(point)
        },
        "metadata": [
          "weight": [
            "time": Utils.localizedValueToJson(weight.time),
            "distance": Utils.localizedValueToJson(weight.walkingDistance)
          ]
        ]
      ]
    }

    let arguments: [String: Any] = [
      "routes": routes
    ]

    result(arguments)
  }
    private func onError(_ error: Error, _ result: @escaping FlutterResult) {
      var errorMessage = "Unknown error"

      if let underlyingError = (error as NSError).userInfo[YRTUnderlyingErrorKey] as? YRTError {
        if underlyingError.isKind(of: YRTNetworkError.self) {
          errorMessage = "Network error"
        } else if underlyingError.isKind(of: YRTRemoteError.self) {
          errorMessage = "Remote server error"
        }
      } else if let msg = (error as NSError).userInfo["message"] {
        errorMessage = msg as! String
      }

      let arguments: [String: Any?] = [
        "error": errorMessage
      ]

      result(arguments)
    }
  }
